lib/xandra/cluster/pool.ex

```diff
@@ -412,21 +412,30 @@ defmodule Xandra.Cluster.Pool do
         data.load_balancing_module.query_plan(lb_state)
       end)
 
-    # Find the first host in the plan for which we have a pool.
-    reply =
+    # Find all connected hosts
+    connected_hosts =
       query_plan
-      |> Stream.map(fn %Host{} = host -> Map.fetch(data.peers, Host.to_peername(host)) end)
-      |> Enum.find_value(_default = {:error, :empty}, fn
-        {:ok, %{pool_pid: pid}} when is_pid(pid) -> {:ok, pid}
-        _other -> nil
+      |> Enum.map(fn %Host{} = host -> Map.fetch(data.peers, Host.to_peername(host)) end)
+      |> Enum.map(fn
+        {:ok, %{pool_pid: pid, host: host}} ->
+          {pid, host}
+
+        _other ->
+          nil
       end)
+      |> Enum.reject(&is_nil/1)
```

Comment:
Please let's do a single pass. Also, you need to check if `pool_pid` is a PID and not `nil`. Refactor to:

```elixir
for host <- query_plan,
    host_info when not is_nil(host_info) <- Map.get(data.peers, Host.to_peername(host),
    %{pool_pid: pid, host: host} = host_info,
    is_pid(pid),
    do: {pid, host}
```